src/table/components/virtualized-table/hooks/use-infinite-scroll-data.ts

```diff
@@ -43,51 +43,143 @@ const createNewRow = (
   };
 };
 
+const isRowMissingData = (rows: Row[], x: number, y: number, width: number) => {
+  const targetRow = rows[y] as Row | undefined;
+
+  if (!targetRow) {
+    return true; // Row is not cached
+  }
+
+  for (let colIndex = x; colIndex < x + width; colIndex++) {
+    if (!targetRow[`col-${colIndex}`]) {
+      return true; // Column is not cached
+    }
+  }
+
+  return false;
+};
+
+const isColumnMissingData = (rows: Row[], x: number, y: number, height: number) => {
+  const targetRows = rows.slice(y, height);
+
+  if (targetRows.length === 0) {
+    return true; // Rows are not cached
+  }
+
+  return targetRows.some((row) => !row[`col-${x}`]);
+};
+
 const useInfiniteScrollData = (
   model: EngineAPI.IGenericObject,
   layout: TableLayout,
   pageInfo: PageInfo,
+  visibleRowCount: number,
+  visibleColumnCount: number,
   columns: Column[]
 ): UseInfiniteScrollData => {
   const [rowsInPage, setRowsInPage] = useState<Row[]>([]);
-  useOnPropsChange(() => setRowsInPage([]), [layout, pageInfo]);
-
-  const loadData: LoadData = useCallback(
-    async (qLeft: number, qTop: number, qWidth: number, qHeight: number) => {
-      const pageRowStartIdx = qTop - pageInfo.page * pageInfo.rowsPerPage;
 
-      const [dataPage] = await model.getHyperCubeData('/qHyperCubeDef', [{ qTop, qLeft, qHeight, qWidth }]);
+  const getDataPages = useCallback(
+    async (pages: EngineAPI.INxPage[]) => model.getHyperCubeData('/qHyperCubeDef', pages),
+    [model]
+  );
 
+  const handleDataPages = useCallback(
+    (dataPages: EngineAPI.INxDataPage[]) =>
       setRowsInPage((prevRows) => {
         const nextRows = [...prevRows];
-        dataPage.qMatrix.forEach((matrixRow, matrixRowIdx: number) => {
-          const { row, pageRowIdx } = createNewRow(
-            matrixRow,
-            nextRows,
-            matrixRowIdx,
-            qTop,
-            qLeft,
-            pageRowStartIdx,
-            columns
-          );
 
-          nextRows[pageRowIdx] = row;
+        dataPages.forEach((dataPage) => {
+          dataPage.qMatrix.forEach((matrixRow, matrixRowIdx) => {
+            const pageRowStartIdx = dataPage.qArea.qTop - pageInfo.page * pageInfo.rowsPerPage;
+            const { row, pageRowIdx } = createRow(
+              nextRows,
+              matrixRow,
+              matrixRowIdx,
+              dataPage.qArea,
+              pageRowStartIdx,
+              columns
+            );
+
+            nextRows[pageRowIdx] = row;
+          });
         });
 
         return nextRows;
-      });
+      }),
+    [pageInfo, columns]
+  );
+
+  // The queue takes a EngineAPI.INxPage object as items and adds them to a queue and
+  // exists to prevent the same page from being fetched more than once.
+  const queue = useGetHyperCubeDataQueue(getDataPages, handleDataPages);
+
+  useOnPropsChange(() => {
+    queue.clear();
+    setRowsInPage([]);
+  }, [layout, pageInfo.page]);
+
+  const loadData: LoadData = useCallback(
```

Comment:
so this runs only in the last useEffect, right? And it listen to changes on `getDataPages` and  `handleDataPages`. Seems a bit overkill to first wrap this with useCallback, then make that useEffect depend `loadData`, instead of just running the logic inside the useEffect. Or does it have something to do that it is async?

Reply:
No we can do as you say, will move it. It so little code that there is no longer any reason to have it's own function.

Reply:
great, helps readability a bit